### ISSUE TITLE
Compilication error

### ISSUE THREAD

**Issue by chasingegg** [guowentian/dynamicppr#1]
Hi, I'm compiling cpu code, but got the following error.
```
../util/SpinLock.h:73:25: error: no match for ‘operator==’ (operand types are ‘const std::atomic_flag’ and ‘int’)
   return (spinlock_.v_) == 1;
```

**Comment by guowentian** [comment]
Have you installed boost correctly ?

**Comment by guowentian** [comment]
Actually boost is not very necessary in this project. You could modify the Makefile in 'cpu' folder like this
+LIBS = -lm -lpthread #-lpapi
+CFLAGS += -O3 -std=c++11 -fcilkplus -lcilkrts -DPTHREAD_LOCK -DPTHREAD_BARRIER
This would use pthread instead of boost.

**Comment by chasingegg** [comment]
OK, thank you.

**Comment by chasingegg** [comment]
sorry, didn't find the **workload** and **encoder** folder?

**Comment by guowentian** [comment]
So sorry that I forgot to include them .... I pushed them into the repository just now....Now you can find them..

**Comment by chasingegg** [comment]
haha, thank you

**Comment by chasingegg** [comment]
Sorry, disturb you again here. For the source vertex "s", you maintain the PPR vector, but you use the parallelized reverse push, so I guess actually the source vertex means the target node, PPR vector here means the PPR from other nodes end at "s", am i right?

**Comment by guowentian** [comment]
@chasingegg You are right. It could be misleading if I call "source vertex", actually it is better be called as "target vertex".

**Comment by chasingegg** [comment]
So nice of you, thank you very much!

**Comment by guowentian** [comment]
@chasingegg you are welcome.

**Comment by chasingegg** [comment]
Sorry, seems I got another problem. The result validation seems not right...

**Comment by guowentian** [comment]
@chasingegg how do you validate the result ?

**Comment by chasingegg** [comment]
Just use the validation part of the code

**Comment by guowentian** [comment]
@chasingegg The code of validation part cannot get the exact solution, unless you give a very small epsilon to it. It is a power iteration method, and it stops when the bound between two iterations are smaller than a thershold, while PPR stops when the residual is smaller than a threshold. These two approximation methods are not comparable actually. In my implementation I just use power iteration to validate ppr result in a high level perspective, but this is not a good way to do it, so I deprecated VALIDATE flag.

**Comment by chasingegg** [comment]
Ok I got it. Just confirm that you will also get the err when using the validation code?  You are right, but the err may not reach to 0.01 because you have set the very small threshold for both two methods.

**Comment by guowentian** [comment]
@chasingegg I remember I don't have err when I set small threshold

**Comment by chasingegg** [comment]
Hello, finally I found out the reason why I get err in the validation part. There is a bug in the validation code. There is no need to perform ```graph->ConstructGraph()``` again when doing the validation. Since when constructing the graph, you don't set edge information to zero, so it got wrong. 
Therefore, a simple way to eliminate the bug is just delete the line ```graph->ConstructGraph()``` in the validation part

**Comment by guowentian** [comment]
@chasingegg what is the edge information you mentioned ?

**Comment by chasingegg** [comment]
For example, in ```GraphVec.h```, ```col_ind[v1].push_back(v2);``` so when constructing graph again in the validation, out_degree can become twice over the original graph since ```col_ind```has values.

**Comment by guowentian** [comment]
@chasingegg Thanks I would check it later. This is the part of static graphs. I haven't checked the local push on static graphs for some time, since my focus here is the dynamic graphs.
